plugins/kinetic-dp/fu-kinetic-dp-secure-aux-isp.c

```diff
@@ -0,0 +1,1256 @@
+			    FWUPD_ERROR_INTERNAL,
+			    "buffer size [%u] is not enough to read DPCD_ISP_REPLY_DATA_REG [%u]",
+			    buf_size,
+			    read_data_len);
+		return FALSE;
+	}
+
+	if (read_data_len > 0) {
+		if (!fu_kinetic_dp_connection_read(self,
+						   DPCD_ADDR_ISP_REPLY_DATA_REG,
+						   buf,
+						   read_data_len,
+						   error)) {
+			g_prefix_error(error, "failed to read DPCD_ISP_REPLY_DATA_REG: ");
+			return FALSE;
+		}
+		*read_len = read_data_len;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_kinetic_dp_secure_aux_isp_write_dpcd_reply_data_reg(FuKineticDpConnection *self,
+						       const guint8 *buf,
+						       guint8 len,
+						       GError **error)
+{
+	gboolean ret = FALSE;
+	gboolean res;
+
+	if (len > DPCD_SIZE_ISP_REPLY_DATA_REG) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "length bigger than DPCD_SIZE_ISP_REPLY_DATA_REG [%u]",
+			    len);
+		return FALSE;
+	}
+
+	res = fu_kinetic_dp_connection_write(self, DPCD_ADDR_ISP_REPLY_DATA_REG, buf, len, error);
+	if (!res) {
+		/* clear reply data length to 0 if failed to write reply data */
+		g_prefix_error(error, "failed to write DPCD_KT_REPLY_DATA_REG: ");
+		len = 0;
+	}
+
+	if (fu_kinetic_dp_connection_write(self,
+					   DPCD_ADDR_ISP_REPLY_LEN_REG,
+					   &len,
+					   DPCD_SIZE_ISP_REPLY_LEN_REG,
+					   error) &&
+	    res)
+		ret = TRUE;
+
+	return ret;
+}
+
+static gboolean
+fu_kinetic_dp_secure_aux_isp_write_mca_oui(FuKineticDpConnection *connection, GError **error)
+{
+	guint8 mca_oui[DPCD_SIZE_IEEE_OUI] = {MCA_OUI_BYTE_0, MCA_OUI_BYTE_1, MCA_OUI_BYTE_2};
+	return fu_kinetic_dp_aux_dpcd_write_oui(connection, mca_oui, error);
+}
+
+static gboolean
+fu_kinetic_dp_secure_aux_isp_enter_code_loading_mode(FuKineticDpConnection *self,
+						     gboolean is_app_mode,
+						     const guint32 code_size,
+						     GError **error)
+{
+	guint8 status;
+	g_debug("enter isp driver loading mode");
+	if (is_app_mode)
+		/* send "DPCD_MCA_CMD_PREPARE_FOR_ISP_MODE" command first
+		 * to make DPCD 514writable */
+		if (!fu_kinetic_dp_secure_aux_isp_send_kt_prop_cmd(self,
+								   KT_DPCD_CMD_PREPARE_FOR_ISP_MODE,
+								   500,
+								   10,
+								   &status,
+								   error))
+			return FALSE;
+
+	/* Update payload size to DPCD reply data reg first */
+	if (!fu_kinetic_dp_secure_aux_isp_write_dpcd_reply_data_reg(self,
+								    (guint8 *)&code_size,
+								    sizeof(code_size),
+								    error))
+		return FALSE;
+
+	if (!fu_kinetic_dp_secure_aux_isp_send_kt_prop_cmd(self,
+							   KT_DPCD_CMD_ENTER_CODE_LOADING_MODE,
+							   500,
+							   10,
+							   &status,
+							   error))
+		return FALSE;
+
+	return TRUE;
+}
+
+/**
+ * fu_kinetic_dp_secure_aux_isp_crc16:
+ * @buf: memory buffer
+ * @bufsz: size of @buf
+ *
+ * Returns the cyclic redundancy check value for the given memory buffer.
+ * This is a proprietary implementation only can be used in Kinetic's
+ * Secure AUX-ISP protocol
```

Comment:
You can't use fu_common_crc16_full?

Reply:
please refer to Jeffery Lin comment on this in PR 3425.  I requote here for convenient.  I think you agreed to let this be static function already.

These days I've studied for CRC-16 things, it's more complicated than what I could imagine.
I tried to find some reference codes to compare the outputs among Kinetic's CRC-16 function, fu_common_crc16(), and other reference implementations(CRC-16, CRC16-CCITT).
I ported the CCITT function from a library called LibCRC.
But I found that the output of fu_common_crc16(init value: 0xFFFF, polynomial: 0x12) is not the same as the output of libcrc.
My references are as below links:
http://srecord.sourceforge.net/crc16-ccitt.html
https://reveng.sourceforge.io/crc-catalogue/16.htm
[LibCRC]
https://www.lammertbies.nl/comm/info/crc-calculation
https://www.libcrc.org/

According to one of above articles, there're many incorrect CRC-16 implementations on the web.
And there're so many variants(parameters) to implement.
May I know how do you implement fu_common_crc16() function and what's your reference?

Back to the PR, I think our CRC-16 implementation just works as a checksum mechanism while transmitting the payload.
I don't think it's a standard CRC16-CCITT implementation(I don't know the history either...).
I can't just replace it with fu_common_crc16() because of the backward-compatible issue in Windows tools released to our customers.
So my first thought is to rename our CRC16 implementation as a proprietary CRC-16 "static" function.
Would it be acceptable?

Thanks a lot.

Reply:
Ahh yes, I forgot about that. All fine then, thanks.